tests/utils/test_schemapi.py

```diff
@@ -871,10 +878,92 @@ def test_chart_validation_errors(chart_func, expected_error_message):
         warnings.filterwarnings("ignore", category=UserWarning)
         chart = chart_func()
     expected_error_message = inspect.cleandoc(expected_error_message)
-    with pytest.raises(SchemaValidationError, match=expected_error_message):
+    with pytest.raises(schemapi.SchemaValidationError, match=expected_error_message):
         chart.to_dict()
 
 
+_SKIP_SLOW_BENCHMARKS: bool = True
+_REPEAT_TIMES = 1000
+
+
+@pytest.mark.parametrize("to_or_from", ["to_dict-validate", "to_dict", "from_dict"])
+@pytest.mark.filterwarnings("ignore:.*:UserWarning")
+@pytest.mark.skipif(
+    _SKIP_SLOW_BENCHMARKS,
+    reason="Should only be run in isolation to test single threaded performance.",
+)
+def test_chart_validation_benchmark(
+    to_or_from: Literal["to_dict-validate", "to_dict", "from_dict"],
+) -> None:
+    """
+    Intended to isolate `Chart.(to|from)_dict.` calls.
+
+    Repeated ``_REPEAT_TIMES`` times, non-parametric:
+    - in an attempt to limit the potential overhead of ``pytest``
+    - but enforce ``1`` thread, like a user-code would be.
+
+    Results
+    -------
+    ```
+    _REPEAT_TIMES = 1000
+    pytest -k test_chart_validation_benchmark  --numprocesses=3 --durations=3 tests
+
+    # Pre-`SchemaBase.from_dict` refactor (3.12.3)
+    108.16s call     tests/utils/test_schemapi.py::test_chart_validation_benchmark[to_dict-validate]
+    84.62s call     tests/utils/test_schemapi.py::test_chart_validation_benchmark[from_dict]
+    66.71s call     tests/utils/test_schemapi.py::test_chart_validation_benchmark[to_dict]
+
+    # Post-`SchemaBase.from_dict` refactor (3.12.3)
+    107.84s call     tests/utils/test_schemapi.py::test_chart_validation_benchmark[to_dict-validate]
+    50.43s call     tests/utils/test_schemapi.py::test_chart_validation_benchmark[from_dict]
+    67.07s call     tests/utils/test_schemapi.py::test_chart_validation_benchmark[to_dict]
+    ```
```

Comment:
If anyone stumbles upon this PR while in draft, this is the headline:

- `SchemaBase.from_dict` is seeing a **1.67x** speedup on this benchmark so far
- Across the whole test suite, `python<3.12` has seen a huge improvement from fewer forced exceptions [`933c045` (#3547)](https://github.com/vega/altair/pull/3547/commits/933c045f54a9987ca6a6c43f22e6a1e33fe842d9)